docs/modules/SDDs/pages/0033-commodore-component-instance-versioning.adoc

```diff
@@ -0,0 +1,91 @@
+It additionally details the options we chose to implement.
+====
+
+== Motivation
+
+With xref:0025-commodore-component-instantiation.adoc[SDD 0025], Commodore can instantiate a component multiple times per cluster if multiple aliases are specified.
+However, all aliases currently use version information from the base component.
+Commodore does not currently support specifying different component versions for each alias.
+
+=== Goals
+
+* Define how different versions for each component instance can be specified and generated.
+
+=== Non-Goals
+
+
+
+== Design Proposal
+
+Commodore will leverage the existing configuration options in `.parameters.components` to retrieve version information for specific instances.
+Instance versions can be specified in `.parameters.components.<instance-name>` and override the value specified in `.parameters.components.<component-name>`.
+Even if all instances provide their own version, a version for the base component (`.parameters.components.<component-name>`) must still be specified.
+
+Component authors must explicitly declare that their component supports multi-versioning.
+In a valid catalog, all used versions of the same component must advertise this support.
+
+Components with multi-versioning support must ensure they create no file conflicts when downloading dependencies.
+In practice, this means components must use the parameter `${_base_directory}` in their compile configuration (`parameters.kapitan.dependencies` and `parameters.kapitan.compile`).
+
+Support for providing different versions of Jsonnet dependencies is currently not implemented.
+Jsonnet dependencies are always provided from the main (non-instantiated) version of the component.
+If a component is to support multi-versioning, Jsonnet dependencies should therefore be compatible within a reasonable span of versions.
+
+Similarly, if the component includes a Jsonnet library that is made available to other components, the version for the library is taken from the main (non-instantiated) version of the component.
+
+=== Implementation Details/Notes/Constraints
+
+We can clone the repository of the base component and create a separate worktree for each instance to check out the correct version.
+
+If the instance version specifies a different URL from the base component, more work is needed: in this case, we clone the changed repository for the instance and create a worktree from that.
+
+This means that if two instances both override the URL to the same value, this repository is cloned twice, which can impact performance.
+However, since this only affects instances which override the repository URL, the impact is tolerable.
+
+=== Risks and Drawbacks
+
+* Since the JSonnet dependencies and libraries are all provided from a single component version, there is a risk for some combinations of versions not compiling properly due to version compatibility issues.
```

Comment:
I'm not sure off the top of my head how stable these types of dependencies are, so I'm not sure how often we'd expect issues with this.
My current thinking is, we can always implement support for properly versioned JSonnet dependencies when we actually need it - since it looks like quite some effort, and we'll still have the same problem with JSonnet libraries.
Happy to hear your opinions.